[IR][LangRef] Add partial reduction add intrinsic

NickGuy-Arm opened the pull request:
Adds the llvm.experimental.partial.reduce.add.* overloaded intrinsic, this intrinsic represents add reductions that result in a narrower vector. In the generic case, this intrinsic is lowered to a chain of add reductions, it is assumed that it will only be emitted for targets that provide their own lowering.

NickGuy-Arm committed:
Add partial reduction add intrinsic

arsenm commented:
What's the benefit to claiming this is a separate operation from the existing reduction intrinsics? Can't you just loosen the type declaration to allow vector types for the result?

paulwalker-arm commented on llvm/docs/LangRef.rst:
```diff
@@ -19209,6 +19209,35 @@ will be on any later loop iteration.
+Overview:
+"""""""""
+
+The '``llvm.vector.experimental.partial.reduce.add.*``' intrinsics do an integer
+``ADD`` reduction of subvectors within a vector, returning each scalar result as
+a lane within a vector. The return type is a vector type with an
+element-type of the vector input and a width a factor of the vector input
+(typically either half or quarter).
```
The intent here is to keep the reduction intrinsics as loose as possible so we don't lock the code generator into a specific ordering.

If there's an option to simply extend the original intrinsics that would be super but I figured it would be easier to move current uses to a newer intrinsic (assuming it leaves the experimental space) than the other way round.

paulwalker-arm commented on llvm/docs/LangRef.rst:
```diff
@@ -19209,6 +19209,35 @@ will be on any later loop iteration.
+
+::
+
+      declare <4 x i32> @llvm.experimental.vector.partial.reduce.add.v2i32.v8i32(<8 x i32> %in)
+      declare <4 x i32> @llvm.experimental.vector.partial.reduce.add.v4i32.v16i32(<16 x i32> %in)
```
Is it best for these intrinsics to take a single operand?  I kind of see them more like binary operators whose input and output operands are less restrictive.

The intent being we're doing a partial reduction where currently LoopVectorize achieves this via an `add` (or other binop) instruction that is too restrictive (i.e. it forces a very specific ordering for how elements are combined)  and so we're introducing a new intrinsic that better reflects what LoopVectorize wants to achieve but without the baggage. In this way existing code can pick a format of the intrinsic that can be mapped directly to an `add` instruction if need be.

paulwalker-arm replied:
To be specific I'm proposing
```
declare <4 x i32> @llvm.experimental.vector.partial.reduce.add.v4i32.v16i32(<4 x i32>,<16 x i32>)
```
whereby the result and first operand types match, but the second operand differs (perhaps with a restriction that is must have the same or more elements).

NickGuy-Arm committed:
Change partial reduction intrinsic to take the accumulator as an operand

NickGuy-Arm committed:
Rename test file

NickGuy-Arm committed:
Fix docs build error

huntergr-arm commented on llvm/test/CodeGen/AArch64/partial-reduction-add.ll:
```diff
@@ -0,0 +1,162 @@
+  ret <vscale x 4 x i32> %partial.reduce
+}
+
+define <vscale x 4 x i32> @partial_reduce_add_quart(<vscale x 4 x i32> %accumulator, <vscale x 16 x i32> %0) #0 {
```
This is reducing into the first 4 elements of the accumulator; it doesn't work correctly with vscale.

huntergr-arm commented on llvm/test/CodeGen/AArch64/partial-reduction-add.ll:
```diff
@@ -0,0 +1,162 @@
+declare i32 @llvm.vector.reduce.add.nxv8i32(<vscale x 8 x i32>) #2
+
+attributes #0 = { "target-features"="+fp-armv8,+fullfp16,+neon,+sve,+sve2,+v8a" }
+attributes #1 = { nocallback nofree nosync nounwind willreturn memory(none) }
```
I think attributes 1 and 2 can be removed entirely, and 0 only really needs +sve2.

arsenm commented:
> What's the benefit to claiming this is a separate operation from the existing reduction intrinsics? Can't you just loosen the type declaration to allow vector types for the result?

ping?

huntergr-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7914,6 +7914,28 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    auto Accumulator = getValue(I.getOperand(0));
+    unsigned ScaleFactor = FullTy.getVectorMinNumElements() / ReducedTy.getVectorMinNumElements();
+
+    for(unsigned i = 0; i < ScaleFactor; i++) {
```
I think the previous design may have been better, since it was clearly just performing the reduction of a single vector value into another (and possibly to a scalar, as @arsenm suggests). Making it a binop as well seems to make it less flexible vs. just having a separate binop afterwards. Maybe I'm missing something though...

paulwalker-arm replied:
The problem with the "having a separate binop" approach is that it constrains optimisation/code generation because that binop requires a very specific ordering for how elements are combined, which is the very problem the partial reduction is solving.

I think folk are stuck in a "how can we use dot instructions" mindset, whilst I'm trying to push for "what is the loosest way reductions can be represented in IR".  To this point, the current suggested langref text for the intrinsic is still too strict because it gives the impression there's a defined order for how the second operand's elements are combined with the first, where there shouldn't be.

paulwalker-arm replied:
@huntergr-arm - Yes, the intent for "same size operands" is to emit a stock binop.  This will effectively match what LoopVectorize does today and thus allow the intrinsic to be used regardless of the target rather than having to implement target specific/controlled paths within the vectorizer.

huntergr-arm replied:
Ok, I was thrown off by the langref description. I guess then I'd like to see the default lowering changed to just extract the subvectors from the second operand and perform a vector add on to the first operand, instead of reducing the subvectors and adding the result to individual lanes. It technically meets the defined semantics (target-defined order of reduction operations), but the current codegen is pretty awful compared to a series of vector adds.

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7914,6 +7914,28 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     setValue(&I, Trunc);
     return;
   }
+  case Intrinsic::experimental_vector_partial_reduce_add: {
```
We need to be careful because I don't think common code exists to type legalise arbitrary INTRINSIC_WO_CHAIN calls (given their nature).  Presumably we'll just follow the precedent set for `get.active.lane.mask` and `cttz.elts` when we add AArch64 specific lowering.

I can't help but think as some point we'll just want to restrict the "same element type" restrict of `VECREDUCE_ADD` to have explicit signed and unsigned versions, like we have for ABDS/ABDU, but I guess we can see how things work out (again much as we are for the intrinsics mentioned before).

paulwalker-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6131,6 +6131,19 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
     }
     break;
   }
+  case Intrinsic::experimental_vector_partial_reduce_add: {
```
See above for my 2c.

NickGuy-Arm committed:
Update LangRef.rst for partial reduction intrinsic

NickGuy-Arm commented:
I've updated the langref, addressing the comments left. However I have not yet modified the codegen. 


> > What's the benefit to claiming this is a separate operation from the existing reduction intrinsics? Can't you just loosen the type declaration to allow vector types for the result?
> 
> ping?

Apologies @arsenm, I missed your message initially.
That would be the ideal result, yes; however at the moment it's useful, until feature parity is achieved, to keep this intrinsic in the experimental namespace. I believe Paul alluded to this further up, but the end goal is to migrate uses of reductions to use this over time, rather than attempting to retrofit vector->vector reductions into the existing reduction intrinsics.

NickGuy-Arm committed:
Update LangRef.rst for partial reduction intrinsic

paulwalker-arm commented on llvm/docs/LangRef.rst:
```diff
@@ -19218,27 +19218,26 @@ This is an overloaded intrinsic.
-input vector down to the number of elements dictated by the result vector, and
-then adds the resulting vector to the accumulator vector. The return type is a
-vector type that matches the type of the accumulator vector.
+second operand vector down to the number of elements dictated by the result
+vector, and then adds the resulting vector to the first operand vector. The
+return type is a vector type that matches the type of the first operand vector.
```
May I suggest 
```
concatenation of its two vector operands down to the number of elements dictated by the result. The result type....
```

davemgreen commented:
Was there an RFC for this anywhere yet to point to the review? The reviewer list looks a bit narrow. Thanks

paulwalker-arm commented:
> Was there an RFC for this anywhere yet to point to the review? The reviewer list looks a bit narrow. Thanks

There is https://discourse.llvm.org/t/rfc-is-a-more-expressive-way-to-represent-reductions-useful/74929, which started the conversation but as you'll see nobody seemed interested and so I said Arm would take the lead.  We should have referenced the PR when available though, which I've rectified.

NickGuy-Arm committed:
Update LangRef.rst for partial reduction intrinsic

NickGuy-Arm committed:
Implement generic lowering for the partial reduction intrinsic

huntergr-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7914,6 +7914,46 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+      Subvectors.push_back(DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, ReducedTy, {OpNode, SourceIndex}));
+    }
+
+    while(Subvectors.size() >= 2) {
```
I think this could be improved a little by using a std::deque instead of a SmallVector. You can push_back the operands as above, but then use a smaller loop like the following to compute the result:

```
while (Subvectors.size() > 1) {
  Subvectors.push_back(DAG.getNode(ISD::ADD, DL, ReducedTy, {Subvectors[0], Subvectors[1]});
  Subvectors.pop_front();
  Subvectors.pop_front();
}
```

Sadly there's no `pop_front_val()` that gives you the value when removing it, but it is shorter.

NickGuy-Arm committed:
Use deque instead of SmallVector for generic lowering

huntergr-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7914,6 +7915,35 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     return;
   }
+  case Intrinsic::experimental_vector_partial_reduce_add: {
+    auto DL = getCurSDLoc();
```
nit: It would be good to remove the 'auto' declarations and use the appropriate named types (SDValue, EVT, int, etc). I think you should already have a variable in scope getCurSDLoc() as well (sdl, from the start of the function).

huntergr-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6131,6 +6131,19 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
+    VectorType *AccTy = cast<VectorType>(Call.getArgOperand(0)->getType());
+    VectorType *VecTy = cast<VectorType>(Call.getArgOperand(1)->getType());
+
+    auto VecWidth = VecTy->getElementCount().getKnownMinValue();
```
nit: more autos.

NickGuy-Arm committed:
Address final comments

NickGuy-Arm committed:
Fix formatting

NickGuy-Arm committed:
Fix formatting using git-clang-format
